crates/flowctl/src/controlplane.rs

```diff
@@ -31,15 +32,30 @@ impl Deref for Client {
     }
 }
 
+#[derive(Deserialize)]
+struct AccessTokenResponse {
+    access_token: String,
+}
+
 /// Creates a new client. **you should instead call `CliContext::controlplane_client(&mut Self)`**, which
 /// will re-use the existing client if possible.
-pub(crate) fn new_client(config: &Config) -> anyhow::Result<Client> {
-    match &config.api {
+pub(crate) async fn new_client(config: &mut Config) -> anyhow::Result<Client> {
```

Comment:
Passing a `&mut Config` here means that the config file will be re-written every time a client is created. If this parameter was a `&mut CliContext`, then it could only call `context.config_mut()` in the case that it actually needs to re-fresh the credential.

My thinking is that long term we should also stop storing the `access_token` in the config file, or maybe even not store it at all. I'm totally on board with _not_ changing that right now, to minimize the scope of the change. But I don't think we should go back to always writing the config file on every invocation.

To give some background on why always writing the config file is undesirable, consider the example where you're running flowctl as part of a Job in a kubernetes cluster. You'd often have that config file defined in a `ConfigMap` and then mount it into the container. But kubernetes updates the mounted files whenever the ConfigMap is updated, which could conflict with flowctl's writes to it. And anything flowctl writes to that file will ultimately be lost, anyway, since the ConfigMap is authoritative. There's other environments where you can really only mount that type of file as read-only. The general takeaway being that if we want flowctl to run in as many different environments as possible, it'll be easier if that file can be read-only.

So I'd like to get to a place where we generally treat that config file as read-only unless the user specifically asks us to update it (by running `auth login` or something). For now, it's totally fine to keep storing the `access_token` there, as long as we're not requiring that it be re-written on every invocation.

Reply:
It's actually not that straightforward, since we can't borrow config of `CliContext` as both mutable and immutable. What you are suggesting would mean that we take a `&mut CliContext` in `new_client`, and then match on `&context.config().api` (immutable borrow), and if conditions make it necessary, we use `context.config_mut().api` to mutate the config (mutable borrow), but compiler will complain about that, unless we want to introduce some cell.

On the other hand, I think this is a hard constraint to maintain when put that way: that we must not modify our config struct throughout the program if the program is non-interactive / running in a specific environment. `flowctl` can only know so much about its environment.

The best I can suggest is introducing an environment variable that prevents writes to config file, and we can use that env in read-only environments if it's really necessary. If it isn't, then I would suggest we just go on as-is.

Reply:
Yeah, you'd need to break it down into two separate borrows. One immutable borrow to check if there's existing credentials and whether they're still valid, and then a separate, mutable borrow to actually update the config if need be.

With regard to not writing to the config file more generally, I think I'd phrase it as more of an axiom about config files rather than a constraint about environments. It's generally bad form to store application managed "state" in a config file because it makes it harder for users to actually use it as a config file, mainly because it breaks lots of assumptions people have about "config" files, especially when considering other environments.

In this case, I think it's totally fine to merge as it is, and we can always revisit the dual borrow later. Perhaps at the same time as we remove the access token from the config file :shrug: